Faster/Leaner genesis bootstrap

nagisa opened the pull request:
When forking mainnet based on [these instructions](https://near.zulipchat.com/#narrow/stream/295302-general/topic/forking.20off.20mainnet/near/341511159), these changes reduce the creation time of the initial rocksdb from multiple days or weeks down to a hour and a half. There are a couple facets to this improvement:

* Process shards in parallel – this mitigates the impact of re-parsing the genesis/records json for every single shard (see https://github.com/near/nearcore/issues/8747 for further improvement opportunities)
* In the old code we had a `O(n^2)` loop where for each block of 300k accounts we would go through the entire set of records from scratch. For mainnet-sized records this is _the_ time sink, given that there are like a good dozen million accounts and each such iteration takes like 5 minutes. Not to mention this code does not achieve the intended purpose of limiting the memory usage at all – as soon as there is a cluster of multiple accounts each with huge amount of state at the same time, the memory usage to store the pending writes would explode. Here I chosen to introduce a context object that limits the number of outstanding writes and processes them in the order they occur in the records file (there is no reason to group the writes by account as far as I can tell.)

nagisa commented:
I kept an eye out for the memory usage throughout the import process and it seems to stay nicely under 9GiB and fairly constant. So that's another improvement this PR achieves.

jakmeier commented on runtime/runtime/src/genesis.rs:
```diff
@@ -71,210 +69,306 @@ impl<'a> StorageComputer<'a> {
+
+    fn modify<R>(&self, cb: impl FnOnce(&mut StateRoot, &mut TrieUpdate) -> R) -> R {
+        let mut guard = self.lock();
+        let (ref mut update_budget, ref mut state_root, ref mut state_update) = &mut *guard;
+        // Try to get some budget from the global pool.
+        if *update_budget == 0 {
+            let writers = self.active_writers.fetch_add(1, atomic::Ordering::SeqCst) + 1;
+            *update_budget = MAX_OUTSTANDING_WRITES / writers;
```
I don't get how this write-budgeting works.

I looks to me like `update_budget == 0` detects when we are the first write operation to be enqueued in our `AutoFlushingTrieUpdate` instance. If so, we lazily allocate some budget from an implicit pool that is shared with all other `AutoFlushingTrieUpdate` instance sharing the same `active_writers` counter. All subsequent writes on this object will use up one of the allocated writes. When the update_budget reaches zero, we flush and remove ourselves from the active writers, which kind of returns the outstanding writes to the pool.

First question: Is my understanding right?


Second question: Does this really limit the total number of outstanding writes? It seems like the first writer gets `MAX_OUTSTANDING_WRITES` / 2, the next gets `MAX_OUTSTANDING_WRITES` / 3, the next `MAX_OUTSTANDING_WRITES` / 4 and so on. (A diverging series, so unlimited outstanding writes as the number of writer grows.)

Third question: This seems a bit complicated... Why not, let's say give a fixed number to each `AutoFlushingTrieUpdate`?

nagisa replied:
> First question: Is my understanding right?

Yes, the understanding here is correct.

> Second question

The upper limit is indeed not `MAX_OUTSTANDING_WRITES`, but the number of outstanding writers is bounded by the number of shards processed in parallel. This should probably be named `TARGET_OUTSTANDING_WRITES`, and a comment what this actually means would help too, sure.

> Third question

As the number of shards increases, a constant limit for `AutoFlushingTrieUpdate` would mean a proportional increase in the memory usage. We could set the limit appropriate for, say, 100 shards, today, but that would mean excessive flushes today, as well as on the long tail when the last shard is being processed. Admittedly I haven’t measured the impact of flushing frequency (due to huge iteration times), so I went with the intuition that keeping it as infrequent as possible is ideal.

jakmeier replied:
Ok, I see, thanks for elaborating.

It's a bit weird to me that we use some arbitrary number to limit the consumption and then go above that nuber with already 3 concurrent `AutoFlushingTrieUpdate` objects. But I did the math quickly and it seems even with 100 concurrent readers we only use 4x the target size. I suppose that limit actually works out pretty nicely when looking into it this way.

Other than that, it makes sense to me now.

Oh and I like the removal of the Mutex :+1:

jakmeier reviewed (commented):
Lovely, lovely stuff! I like how you got rid of the `Rc` in `TrieUpdate`, which has been annoying to me a couple of times when parallelizing storage operations. I never really considered removing the `Rc`, thinking it surely is there for a good reason. But I can't find a good reason now, so I guess we can remove it and be happy.

But I did comment on a few things I didn't understand, please take a look.

nagisa commented on runtime/runtime/src/genesis.rs:
```diff
@@ -71,210 +69,306 @@ impl<'a> StorageComputer<'a> {
+    }
+
+    fn lock(&self) -> std::sync::MutexGuard<State> {
+        self.state.lock().unwrap_or_else(|g| g.into_inner())
```
Mutex poisoning is one of those things whose utility is a subject of heated debate in the Rust community. See https://blog.rust-lang.org/2020/12/11/lock-poisoning-survey.html for example.

For nearcore specifically, `--panic=abort`, so poisoned mutexes are an impossible state in the first place. But otherwise the choice is quite arbitrary on my part here.

jakmeier replied:
I see, thanks for the explanation.

nagisa commented:
I simplified the logic for outstanding write accounting and removed the mutex as it was only useful to parallelize a loop that does not really parallelize much anyhow.

Longarithm commented on runtime/runtime/src/genesis.rs:
```diff
@@ -71,111 +68,211 @@ impl<'a> StorageComputer<'a> {
-                StateRecord::Account { account_id, account } => {
-                    set_account(&mut state_update, account_id, &account);
-                }
+            match record {
```
nit: I think we can move `storage.modify` outside the match for all cases except `postponed_receipts` and it would make logic slightly simpler.

nagisa replied:
Not all of the branches modify the storage, though. Back when `modify` locked a mutex this current approach actually was the only sensible option. In that context handling a postponed or delayed receipt would result in an unnecessary locks being taken. Given that future modifications may still end up using a `Mutex` for one reason or another (account storage computation is pretty slow, for example) I think the minimal scoping of this approach is ideal.

Longarithm replied:
I see! Currently number of items corresponding to congested receipts is negligible, but who knows how big congestion will be in the future.

jakmeier reviewed (approved):
My concerns have been addressed, so I will give an approval from my side.

Longarithm commented on core/store/src/trie/update.rs:
```diff
@@ -126,7 +122,9 @@ impl TrieUpdate {
     }
 
-    pub fn finalize(self) -> Result<(TrieChanges, Vec<RawStateChangesWithTrieKey>), StorageError> {
+    pub fn finalize(
```
As we change it - can we add a comment saying that it prepares state changes to apply to storage in the future and returns underlying `Trie` back?
Maybe it is obvious for most people because it is called `finalize` and takes `self`, but it never was the case for me.

nagisa commented on runtime/runtime/src/genesis.rs:
```diff
@@ -71,111 +68,211 @@ impl<'a> StorageComputer<'a> {
-        *current_state_root = tries.apply_all(&trie_changes, shard_uid, &mut store_update);
+        // Temporarily store a trie update with the "wrong" state root. This will be replaced
+        // shortly as soon as we know the new state root.
+        let old_state_update = std::mem::replace(
```
This is not that different from using an option – conceptually both `None` and this update with an old state root are logically sentinel values.

I don’t particularly mind adding Option though, this specific field is pretty well abstracted by the API, and performance implications of checking the `Option` aren’t particularly significant.

Longarithm reviewed (approved):
LGTM!

marcelo-gonzalez commented:
fyi tested this on a recentish mainnet snapshot and checked that the state roots match up before & after this PR
